cdap/resource_streaming_program_run.go

```diff
@@ -0,0 +1,335 @@
+	argsObj[FAUX_RUN_ID] = runId.String()
+
+	b, err := json.Marshal(argsObj)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, startAddr, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+
+	if _, err := httpCall(config.httpClient, req); err != nil {
+		return err
+	}
+
+	// Poll until actually reaches RUNNING state.
+	return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		isRunning, err := isFauxRunIdRunningYet(config, runsAddr, runId.String())
+		if err != nil {
+			return resource.RetryableError(err)
+		}
+		if isRunning {
+			d.SetId(runId.String())
+			return nil
+		}
+		time.Sleep(10 * time.Second) // avoid spamming retries
+		return resource.RetryableError(fmt.Errorf("still waiting for program run with faux run id: %s which is in an initializing state", runId.String()))
+	})
+}
+
+//TODO ?
+func resourceStreamingProgramRunRead(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+type RuntimeArgs struct {
+	FauxRunId string `json:"__FAUX_RUN_ID__"`
+}
+
+type RuntimeProperties struct {
+	RuntimeArgs json.RawMessage `json:"runtimeArgs"`
+}
+
+// These are the only keys we need
+type Run struct {
+	RunId      string `json:"runid"`
+	Status     string `json:"status"`
+	Properties RuntimeProperties
+}
+
+// Checks if there is a running run for the terraform faux run id
+// raises error if the program is not in an initializing state (e.g. it failed or was killed in the ui)
+func isFauxRunIdRunningYet(config *Config, runsAddr string, runId string) (bool, error) {
+	s, err := getProgramRunStatusByFauxId(config, runsAddr, runId)
+	if err != nil {
+		return false, err
+	}
+
+	if s == "RUNNING" {
+		return true, nil
+	}
+	if !INITIALIZING_PROGRAM_RUN_STATUSES[s] {
+		return false, fmt.Errorf("Program not Running or Initializing, in state: %s", s)
+	}
+	return false, nil
+
+}
+
+func getProgramRunStatusByFauxId(config *Config, runsAddr string, runId string) (s string, err error) {
+	r, err := getRunByFauxId(config, runsAddr, runId)
+	if err != nil {
+		return
+	}
+
+	return r.Status, nil
+}
+
+// TODO optimization: we call this function often when we could probably get the real runid once and cache it.
+// This would avoid redoing this loop everytime to get the same result. probably inconsequential unless there are many runs of this program
+func getRunByFauxId(config *Config, runsAddr string, runId string) (r Run, err error) {
+	req, err := http.NewRequest(http.MethodGet, runsAddr, nil)
+	if err != nil {
+		return
+	}
+
+	b, err := httpCall(config.httpClient, req)
+	if err != nil {
+		return
+	}
+
+	var runs []Run
+	var args RuntimeArgs
+
+	err = json.Unmarshal(b, &runs)
+	if err != nil {
+		return
+	}
+
+	for _, r := range runs {
+		// Unescaping runtime Args which are stored as an escaped JSON string.
+		unquotedArgs, _ := strconv.Unquote(string(r.Properties.RuntimeArgs))
```

Comment:
check err

Reply:
Also, given how RuntimeArgs is used, I don't think you need type RawMessage, just define it as a string to begin with and put a comment in the struct that the API returns it in a quoted string.

Reply:
If I make this string unuote is throwing an invalid syntax error (I think because unmarshal is dropping the outermost quotes from the string (note I think that the escaping of the inner quotes is dropped in the formatting of the error so the string itself is not valid json))

```
invalid syntax. runtime arguments was: {"cardiology_ocn_sub":"gd-phi-adp-hl7v2-cardiology_ocn","logical.start.time":"1583436617208","system.profile.name":"adp-aintegration-hl7-stream-ingest-us-central1-n","restricted_project_id":"gd-mps-adp-arestricted-n-de98","__FAUX_RUN_ID__":"9b66833a-a453-4e05-b414-038568e13fe3","hl7_routed_sub_prefix":"gd-phi-adp-hl7v2-","gcs_cardiology_uploads":"gd-josp-phi-cardiology-uploads-adp-us-n","gcs_hl7_backfill":"gd-josp-phi-hl7-backfill-adp-us-n"}
```

If it's all the same to you I'd like to leave it as is.

Reply:
Ah ok then this makes sense.

My only feedback would be to prefer using the type []byte instead of RawMessage. RawMessage itself is just a []byte: https://godoc.corp.google.com/pkg/encoding/json#RawMessage but its mainly used when you want to unmarshal it again.

Okay, so my go readability ocd is not going to let me get away without mentioning this

(optional)

A really clean solution here would be writing a custom unmarshal func like this:

```
type runtimePropertiess struct {
	RuntimeArgs runtimeArgs `json:"runtimeArgs"` // renamed from parsedRuntimeArgs
}

func (ra *RuntimeArgs) UnmarshalJSON(data []byte) error {
     unquoted, err := strconv.Unquote(string(data))
      if err != nil {
         return nil, fmt.Errorf("failed to escape runtime arguments %v: %v", string(data), err)
      }

     // Use alias to avoid infinite recursion: https://stackoverflow.com/q/52433467
     type alias runtimeArgs
     var a alias
     if err := json.Unmarshal([]byte(unquoted), &a); err != nil {
         return err
    }
    *ra = a
}
```

Then the caller just becomes:
```
for _, r := range runs {
  if runId == args.FauxRunID {
     ....
   }
}
```